File: Python/Backtracking/optimizedCombination.py
def optimizedCombine(n, k):
    res = []

    def helper(start, curr):
        if len(curr) == k:
            res.append(curr[:])  # copy list
            return

        need = k - len(curr)

        # pruning để tránh duyệt dư
        for j in range(start, n - (need - 1) + 1):
            curr.append(j)
            helper(j + 1, curr)
            curr.pop()

    helper(1, [])
    return res

File: TypeScript/Backtracking/optimizedCombination.ts
function optimizedCombine(n: number, k: number): number[][] {
    const res: number[][] = [];

    function helper(start: number, curr: number[]) {
        if (curr.length === k) {
            res.push([...curr]); // clone array
            return;
        }

        const need = k - curr.length;

        for (let j = start; j <= n - (need - 1); j++) {
            curr.push(j);
            helper(j + 1, curr);
            curr.pop();
        }
    }

    helper(1, []);
    return res;
}